### ISSUE TITLE
[UX] Throw a warning if max_input_vars is 1000 (default PHP setting)

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#3824]
**Description of the bug**

I just got bitten by the problem where PHP's `max_post_vars` is set to 1000 by default. This is a really frustrating bug because there are no errors or warnings, things just act funny when not all the values in a node form (for example) are passed through `$_POST`.

For backdrop sites, we should warn people when this value hasn't been adjusted appropriately. I was thinking of putting this in the status report, perhaps with a recommended value in the `ini_set()` section of settings.php

**Comment by klonos** [comment]
Related Drupal core issues:

https://www.drupal.org/project/drupal/issues/1203766 (specific to the permissions page)
https://www.drupal.org/project/drupal/issues/1565704

Additionally, the same problem seems to be happening in many contrib projects:

https://www.drupal.org/project/cat/issues/2403993
https://www.drupal.org/project/features/issues/2754005
https://www.drupal.org/project/menu_editor/issues/2950256
https://www.drupal.org/project/gathercontent/issues/2948437
...

**Comment by jenlampton** [comment]
This one got me again today...  PR on the way.

**Comment by jenlampton** [comment]
Interesting, it looks like we can fix this in .htaccess for all sites, and we already have for those running PHP5. See [the related drupal issue](https://www.drupal.org/project/drupal/issues/2825895).

This PR includes a fix for PHP7, as well as the warning in the status report for those not using Apache.

I could use some help with the language, but here's what I'm starting with...

Name of the setting:
`PHP form input max`

Error message when not properly set:
`PHP limits the number of form values to 1,000 items.`

Description when not properly set:
`When set too low, PHP's max_input_vars setting can cause silent failures on large forms. Use caution when using the permissions page, ordering large menus, arraging large vocabularies, or editing content with lots of fields.`

Message when properly set:
`PHP limits the number of form values to 10,007 items. This number should be sufficient.`

**Comment by klonos** [comment]
> Name of the setting:
>`PHP form input max`

Perhaps `PHP form input limit`?

Typo on `arraging` in the description.

I like the rest of the wording 👍

**Comment by klonos** [comment]
Thanks for this @jenlampton 👍. Please check my comments in the PR.

...in the meantime, I have closed #3587 as a duplicate of this issue here 😉

**Comment by jenlampton** [comment]
I don't think there is a min version? https://backdropcms.org/requirements

**Comment by jenlampton** [comment]
There are ifs for php modules in .htaccess, maybe that's what you were
thinking of?

**Comment by bradbulger** [comment]
Yes. But if I am reading PR: backdrop/backdrop#2955 right it looks like the simple solution, a mod_php5 block and a mod_php7 block, which should be fine.

**Comment by quicksketch** [comment]
Looks good, two items of feedback:

- Can we put `<IfModule mod_php7.c>` first? We'll eventually remove the PHP 5 block, it should be given secondary treatment.
- The addition to `hook_requirements()` is great, but we should provide a link to documentation on how to increase the limit.

**Comment by bradbulger** [comment]
When running updatedb from drush, the system module warning about max_input_vars is triggered, because command line PHP doesn't read .htaccess settings. It will show every time updatedb is run. I don't know how much that matters. Could this make a difference to tests run from the command line, for instance?

**Comment by klonos** [comment]
This works as expected. Still NW in order to move the php7 block in `.htaccess` before the php5 block + some wording suggestions.

As for the documentation, should that go under the PHP section in https://backdropcms.org/requirements ? We've set the default value in `.htaccess` to `10007`, but what is the minimum safe/recommended in case people cannot set it as high as that? 5000?

**Comment by jenlampton** [comment]
> but we should provide a link to documentation on how to increase the limit.

I'm working on a documentation page at https://api.backdropcms.org/max-input-vars

**Comment by klonos** [comment]
Documentation looks good 👍 ...thanks @jenlampton

**Comment by quicksketch** [comment]
> I'm working on a documentation page at https://api.backdropcms.org/max-input-vars

Thanks @jenlampton! One thing we need to check: I don't think setting `max_input_vars` in settings.php works. Backdrop sets a number of INI values in the bootstrap process, but `max_input_vars` isn't one of them because it has to be set before PHP executes anything at all. My understanding is that this is because the global `$_POST` variable is set based on this value, increasing it after PHP has already started can't restore the truncated values to `$_POST`. So we should either delete that section of the documentation, or perhaps better yet explain why doing it in settings.php won't work.

**Comment by quicksketch** [comment]
I pushed an update to the PR with the changes @klonos and I recommended: https://github.com/backdrop/backdrop/pull/2955

Docs still need to be tweaked per above comment.

**Comment by jenlampton** [comment]
I updated the docs page with how to set the value in .htaccess, and in httpd.conf, and I removed the bit on settings.php (and added why)

**Comment by jenlampton** [comment]
The grammar isn't quite right here: `When set too low, the <code>max_input_vars</code> PHP setting can cause silent failures on forms with a large number of input fields, such as the permissions page or structuring large menus`.

To simplify: `X can cause Y on forms such as structuring large menus.`

The problem is the addition of `structuring large menus`. I wonder if that adds much value here? 

How about we shorten it to: `When set too low, the <code>max_input_vars</code> PHP setting can cause silent failures on any form with a large number of input fields, such as the permissions page.`

I changed `on forms with` to `on any form with` to indicate that there are many of them, but then removed the problematic addition of the second example.

Another alternative might be something like `When set too low, the <code>max_input_vars</code> PHP setting can cause silent failures on forms with a large number of input fields. Examples include the permissions page, the 'Manage links' page for large menus, and the 'List terms' page for large vocabularies.`

This seemed to verbose to me, but feedback welcome :)

**Comment by jenlampton** [comment]
I would mark this RTBC but I'd like someone else to review the final language for the status report :)

**Comment by olafgrabienski** [comment]
In my opinion there are some issues with this PR but I'm not sure how to address them:

- Some people may never have an issue with a value of `1000` but they'd see the big red error all the time. Does the message has to be an error, or would a warning be sufficient?
- If we see a low `max_input_vars` value as a requirement error, should we also mention the requirement on https://backdropcms.org/requirements?
- Regarding documentation on the API page: Will one of the approaches actually increase `max_input_vars`, even on various shared hosting services? If not sure, maybe add something like "If that doesn't help, ask your hosting provider".

**Comment by quicksketch** [comment]
> Regarding documentation on the API page: Will one of the approaches actually increase max_input_vars, even on various shared hosting services? If not sure, maybe add something like "If that doesn't help, ask your hosting provider".

Almost all Apache-based hosting providers respect `.htaccess` settings, so for most sites Backdrop will work with the out of box .htaccess file. I've also seen a lot of shared hosting providers allow you to create a custom `php.ini` file in your home directory that could also be used.

**Comment by indigoxela** [comment]
> Almost all Apache-based hosting providers respect .htaccess settings

That's right.

But one remark of @olafgrabienski is also true:

> Some people may never have an issue with a value of 1000

I only had problems once with D7 and that had really huge ajax based forms.
So there could be rare (?) cases when the big fat red error message appears, but the admin couldn't do anything about it. And they wouldn't have any trouble with the default value at all.

Maybe one possible solution: allow to override this in settings.php? Like, for example, is possible in D7 with the (often wrong) "HTTP request status"? Or is this too much effort?

**Comment by jenlampton** [comment]
What if it was a warning instead of an error?

**Comment by findlabnet** [comment]
Even warning seems like call to action which is unnecessary for most users. 
Point in the installation documentation (or some kind of FAQ) can be useful instead.

**Comment by olafgrabienski** [comment]
Some more real life feedback: I've encountered the issue only *once* so far, caused by a webform with many (but really many) conditionals.

**Comment by jenlampton** [comment]
The problem is extra frustrartng since the failure is completely silent.
Most people won't even know their site isn't submitting all the values it
needs until things are horribly corrupted, or lots of data has been lost.
It's a really terrible problem.

Since the number of people who aren't using apache is very small, and the
number of people who won't be able to change the setting on their own is a
fraction of that fraction, showing these people a warning shouldn't be a
cause for concern.

**Comment by findlabnet** [comment]
So, I must change the setting which is not relevant for me just to not see this annoying warning sign?

**Comment by jenlampton** [comment]
You might find it annoying, but it is an important warning for everyone.

> the setting which is not relevant for me

Just because it's not relevant for a particular site right now does not mean that it will always not be relevant. One day your site might grow, or you may need to include a long and complicated survey. 

If you are comfortable with taking the risk of loosing data and/or having things like your permissions, menu, taxonomy terms and webforms becoming corrupted, then you do not need to make the change. But, when your data is lost, at least we warned you that it would happen by providing a friendly message on the status report :)

**Comment by findlabnet** [comment]
> Information icon with link to the correct documentation (or some kind of FAQ) can be useful instead.

This is not enough?

**Comment by jenlampton** [comment]
No. Without a warning people won't realize that they are in danger without configuring their server properly. They need to be warned of the dangers, so that they can decide if taking the risk is acceptable to them.

**Comment by findlabnet** [comment]
Sometime people can be not so "generic" as assumed. Sure, feel free to ignore this opinion.

**Comment by indigoxela** [comment]
> The problem is extra frustrarting since the failure is completely silent.

I understand that and am all for raising the `max_input_vars` wherever we can (apache).

But I also understand the concern regarding "Something's calling for immediate action, but no action is needed at all".
Showing an error to all nginx and lighty users and all apache users who didn't update their htaccess file seems overdone.

Not the variable is causing disagreement here, it's the "REQUIREMENT_ERROR" state. Problems with reaching max_input_vars don't happen that often. And a value higher than 1000 for max_input_vars is not actually a requirement for *all* Backdrop installs. It's just necessary for *some* - and for sure less than 20%.

I'm curious: the PR sets the value to `10007`. Does this have special meaning, or is that "7" just a typo?
And ten times the default value is possibly a little too high.

**Comment by indigoxela** [comment]
Oops, just realized another glitch...

` <IfModule mod_php7.c>`

These settings, of course, just apply to _apache with mod_php_, but does _not_ apply to apache with fcgi nor apache with php-fpm. The latter are the most common in shared webhosting.

This also means, only a small part of users will actually benefit from that htaccess entry, all the others would get the requirement error. Let's think this over.

**Comment by jenlampton** [comment]
> Showing an error to ...

Downgrading the error to a warning sounds like the way to go. I'm just not okay with not telling people anything about the problem, especially when it's so easy to help people prevent it. (Even on shared hosts most people do still have the ability to update `php.ini`)

I've updated the PR + screenshot in the Original Post to remove any confusion about the warning/error level. 

> <IfModule mod_php7.c>

We can chat about if/how to address other apache/php configurations in another issue if necessary, but that shouldn't block progress here, as this PR uses the same approach as all the other `.htaccess` rules we are using. :) 

Another thought...  @quicksketch mentioned the idea of a dismissible warning. I did some quick research and found that we'd already discussed [dismissible `backdrop_set_message()`s](https://github.com/backdrop/backdrop-issues/issues/4571) and  [dismissible dashboard blocks](https://github.com/backdrop/backdrop-issues/issues/3996), but not [dismissible status report entries](https://github.com/backdrop/backdrop-issues/issues/4617)

I have created a new issue where we can discuss that further :)

**Comment by olafgrabienski** [comment]
Sounds good to me! btw similarly suggested in https://github.com/backdrop/backdrop-issues/issues/3824#issuecomment-686969699 😄 

> Does the message has to be an error, or would a warning be sufficient?
(As far as I know, we don't have dismissible messages so far.)

**Comment by jenlampton** [comment]
> Sounds good to me!

That sounds like RTBC? tagging for one final code review before merge.

**Comment by quicksketch** [comment]
Let's bump this issue to 1.17.1 while we collect additional feedback. I don't see any reason to rush it.

**Comment by BWPanda** [comment]
I keep coming across this issue (as in this page on GitHub, not the problem on Backdrop sites) because of the RTBC label. I finally took the time to read through the issue today to see what's going on here. I have a few comments/questions (though probably nothing really substantial):

- I never knew anything about this PHP setting before now. I just checked one of my Backdrop sites (which is part of my multisite setup, so they all share the same PHP configuration) and it's set to `1,000`. The same is true for my Drupal 7 multisite setup. That's 20 live websites, and I've never had any problems with this setting on any of them...

- The current recommendation as per the PR is to display a warning when this value is set to `1,000` or lower. I noticed on my sandbox site that I have a similar warning on the Status Report page that I haven't set the trusted host setting. It says "It is highly recommended that you configure this to protect your site." If a high recommendation for security reasons is displayed as a warning, is this PHP setting really on the same level as that?

- Finally, a code review question: in the PR it adds the following settings/values to the `.htaccess` file for PHP7:
  - php_flag session.auto_start
  - php_value mbstring.http_input
  - php_value mbstring.http_output
  - php_flag mbstring.encoding_translation
  
  I haven't seen any mention of these in this issue... Why are they included?

**Comment by BWPanda** [comment]
Changing labels based on @quicksketch's comment above.

**Comment by jenlampton** [comment]
> That's 20 live websites, and I've never had any problems with this setting on any of them...

It's also possible that you've never noticed the problem, since the failure is completely silent. I've had editors struggle with both node edit forms and taxonomy re-arrange forms for a while before realizing that something was wrong, and asking for help. 

I also have about 20 live backdrop sites, and have run into the problem on more than 25% of them (though on different forms). It's also worth noting that most of my sites are not terribly complex. Two have large taxonomies, two have a multi-value field on a node type, one has a really deep menu, and one has a really complex webform.  

> If a high recommendation for security reasons is displayed as a warning, is this PHP setting really on the same level as that?

I'd argue that this warning is more important. The "recommendation" is a decision you can make about how you choose to run your site, but nothing will be broken if you choose not to make the change. In this case, it's just a matter of time before your site stops functioning normally.

> I haven't seen any mention of these in this issue... Why are they included?

Those lines are there in the `.htaccess` for PHP 5, but had been forgotten for PHP 7. At this point it might be smart to move those changes out of this PR and into a separate issue. (At the time this was only a critical bug fix change to `.htaccess` -- no warning on the status report - and it seemed faster to get both changes done at once.) Tah Dah! https://github.com/backdrop/backdrop-issues/issues/4752

**Comment by klonos** [comment]
FTR, In today's dev meeting, this idea was thrown around: check the actual number of input vars with each form submission, and then compare that to the php setting; if the number exceeds the php setting, then trow a useful error for people to know what's going on if their forms are not being submitted, or if they are partially saved.

**Comment by jenlampton** [comment]
ok, I tried adding a message when the form was submitted: 
<img width="1254" alt="Screen Shot 2020-11-12 at 3 04 01 PM" src="https://user-images.githubusercontent.com/397895/99007068-5155d100-24f8-11eb-8a77-6be614fe3e4c.png">

It turns out we can also add a message when the form is rendered, before it's submitted. Trying that now too...

**Comment by BWPanda** [comment]
As @olafgrabienski mentioned in the meeting video, Webform has a check for this: [webform_input_vars_check(](https://github.com/backdrop-contrib/webform/blob/1.x-1.x/webform.module#L5395)

Might be worth looking into that...

**Comment by BWPanda** [comment]
I like the idea of the form-specific warning when this is/will be an issue (as opposed to all the time on the Status Report page).

**Comment by klonos** [comment]
Oooh, I also like where this is going!! ...and yes, warning people before they submit the form would be preferable 👍 

Wondering if we could ship our .htaccess file with the changes, but all that commented out. Then we can have people comment them out instead of having to manually create these entries themselves.

**Comment by klonos** [comment]
...one thing to consider if we implement this when the form is being rendered is fields added via AJAX, after the initial rendering of the form (like "add another" in multi-cardinality fields, and paragraphs).

**Comment by klonos** [comment]
...webform has `webform_input_vars_check()` as @BWPanda mentioned, but also [webform_pre_render_input_vars()](https://github.com/backdrop-contrib/webform/blob/1.x-1.x/webform.module#L5425). So they seem to be checking things after form render + after form submit.

**Comment by jenlampton** [comment]
> Wondering if we could ship our .htaccess file with the changes, but all that commented out. Then we can have people simply enable/tweak these entries, instead of having to manually create them from scratch themselves.

We should definitely not have them commented out. There's no harm in fixing this bug for people when we can -- the only debate is what to do for people when we can't fix the bug via `.htaccess`. 

I also really like this idea of throwing a warning on a form we know will be problematic. It will catch the same issue when large/complex sites exceed the `10,007` limit, as well as for all the simple ones that aren't using `mod_php`, and hit the `1000` limit :) 

> and yes, warning people before they submit the form would be preferable

I like the dedicated function approach webform is using, then we can use the same code in _pre_render and _process. I'll give that a go.

**Comment by jenlampton** [comment]
> ...one thing to consider if we implement this when the form is being rendered is fields added via AJAX, after the initial rendering of the form (like "add another" in multi-cardinality fields, and paragraphs).

Webform handles this by adding a 5% buffer:

```php
if ($count > $limit * 0.95) {
```

I've copied some of the webform in the latest PR.

I used the `_count_terminals()` function to estimate number of elements in a given form array, but webform used it's `_count_terminals()` function only to count values in `$_POST`. To count possible elements in a form before it was rendered, webform used `webform_count_input_vars()` instead. Since the code in the PR is not counting $_post directly we might want to switch to something more similar to `webform_count_input_vars()` but it was giving me trouble, and I wanted to have a PR up that people could play with in the sandbox.

**Comment by klonos** [comment]
>> Wondering if we could ship our .htaccess file with the changes, but all that commented out. ...
>
> We should definitely not have them commented out. There's no harm in fixing this bug for people when we can...

Fair enough 👍 

I had a quick look in the code changes, and it looks good overall 👍 ...I like the fact that there's a different message (with less technical lingo) shown to people that do not have permission to manage the site 👍 

>Webform handles this by adding a 5% buffer

Like that too 👍 

> I also tried to do two separate checks, one that was a warning (on form build) and one that was an error (on form submit) but we get both messages on submit because the form is also re-rendered when it is submitted ...

Bummer 😞 ...it would be perfect if we figured that one out, and showed warning on render vs. error on submit.

**Comment by klonos** [comment]
...can someone explain to me what "terminals" are in this context? 😅

**Comment by BWPanda** [comment]
I think the `system_requirements()` code should be removed in favour of the new form checking (or only displayed if a form first triggers this). Also, the new `_form_check_input_vars()` function accepts only one parameter, but two are sent in `backdrop_process_form()`.

**Comment by jenlampton** [comment]
> I think the system_requirements() code should be removed

I'm still on the fence about removing the warning...  I think it would be most helpful as a reminder after install. I'd much rather prevent the problem than have to experience it before learning that it needs to be fixed.

If we do decide to remove the warning, we should at least keep the line that shows the current `max_input_vars` setting.

**Comment by indigoxela** [comment]
I definitely like the direction in this discussion! :+1: I still think we should _not_ worry all site admins with a warning on the status page for a problem they might not understand nor have.

Throwing a warning as the problem actually arises (with too large forms) seems more sophisticated.

@jenlampton I'm pretty sure, code in function _form_count_terminals() could be written in a more readable way. :smirk:

**Comment by indigoxela** [comment]
It seems important to me that we are clear in the issue title and description, that the .htaccess update can only fix it for apache + **mod_php**. The change does not fix it for apache + php-fpm, apache + fastcgi/fcgi, nginx, ... other stuff. I know, I repeat myself :wink: 

If we're able to catch problems in forms (on load and/or submit), an update of the issue description would be helpful, too.

**Comment by jenlampton** [comment]
> It seems important to me that we are clear in the issue title and description,

I've removed `Set max_input_vars for mod_php7 in .htaccess` from the issue title since that has already been handled in `Missing .htaccess rules for PHP 7` #4752

I've also added the more recent screenshot of the form submission failure to the description.

**Comment by BWPanda** [comment]
Yeah, I'm seeing this as being very similar to the trusted host setting, and am coming around to the idea of displaying a yellow warning about this on the Status Report page.

I think a new, separate issue could be opened to discuss the possibility of making these warnings dismissable somehow if people decide they're not relevent to their site and don't want to keep seeing them...

**Comment by klonos** [comment]
@jenlampton the PR has conflicts in `.htaccess` that prevent the sandbox from spinning up.

**Comment by jenlampton** [comment]
I rebased off 1.x and the sandbox should be working again now. (This issue still needs work on the code for counting form elements, and maybe also on message handling on forms)

**Comment by philsward** [comment]
I was going to suggest an "information" icon, but that's already been suggested.

First and foremost: I personally don't care one way or the other how the status message looks. I'll just be happy to have it.

>No. Without a warning people won't realize that there is a risk associated with not configuring their server properly. They need to be warned of the danger, so that they can decide if taking the risk is acceptable to them.

I agree. However, the problem rate is very small across the board. I've had one site 13ish years cause me this problem on more than one occasion. So, warning 95% of sites about an issue they'll likely never run into might be a bit of overkill. If anything, can we do a lookup query on the DB and detect how many "fields" are housed within an entity and if it's a lot, change the status to a warning? (More ideas below)

I think an information line is good enough and here's why:
A new person to Backdrop, will run across that line and read it and ignore it. Over time, they rescan that line and continue to ignore it. More time passes, they learn what they're doing, they land a big site and build out that site with lots of input forms, only to have them disappear on them. Then they remember "Oh yeah, didn't the status page warn me about this? Let's go look at it again."

I think that by having the info there, people will see it and know where to find it if they run into it, however the percentage of people who will actually run into it, are fairly small and the ones that do, have a pretty good handle on knowing what's going on to figure out where to find the fix.

Another idea is to change the status based on the number of vars available.

Example:
250-499 = Stop! Your max_input_vars are set to 300 which is dangerously low and may cause lots of problems with forms. We highly recommend raising them to a minimum of 1000. Here's how to fix it [link]

500-999 = Warning! Your max_input_vars are set to 750 which is tolerable, but may cause problems. We suggest raising them to a minimum of 1000. Here's how to fix it [link]

1000-1499 = Info. Your max_input_vars are set to 1200 which is above the recommended minimum of 1000. If you build a big site with lots of forms, you might need to raise this. Here's some details [link]

1500+ = Success! Your max_input_vars are set to 1600 which is way more than the recommended amount of 1000. Want to learn more? [link]

---

IMO if @jenlampton wants a warning, I say make it a warning and move on. Otherwise I think the best route is to change the status based on the number of vars available. If it's below 1000, show the warning and if it's above 1000 show it as information.

Isn't a warning supposed to make you pay attention but not necessarily take an action?

Are we overthinking it?

**Comment by jenlampton** [comment]
> Otherwise I think the best route is to change the status based on the number of vars available. If it's below 1000, show the warning and if it's above 1000 show it as information. If 1000 isn't high enough, raise it up.

This suggestion is what is already in the current PR.

**Comment by philsward** [comment]
> Some data may have been lost.

Lost is scary.

What about:

It's possible not all new data was saved on this page.

Or something along the lines of "not saved".

"Lost" for me says "wait... what?!??" *Panic*

"Not saved" for me says "Aww common!" *Annoyed*

Thinking about it, lost might be OK, but we need to specify that it's newly entered data. When I first read that my mind went to dark places thinking existing data (from previous saves) might be gone too.

**Comment by jenlampton** [comment]
> thinking existing data (from previous saves) might be gone too

If the fields are missing from the form in POST that cause old data to be overwritten with "nothing" when the item is saved.

**Comment by philsward** [comment]
Ooohh... I stand corrected

I don't remember having that happen to me.

The bold warning you're advocating makes a lot more sense.

**Comment by klonos** [comment]
> I moved the element counting inside form_builder() because that function already iterated form arrays.

Noice 👍 

> I could use some feedback on the messaging.

The warning should be shown after the form loads. When the error is shown (after the form has been submitted), the warning should be hidden. Otherwise it seems like unnecessary duplication.

Also, showing both a success and an error message may be confusing.

I think that all users should see this same error after submitting the form (no success message):

> Changes have been saved, but some data may have been lost. [ERROR]

Additionally, a single warning should _always_ be shown on potentially problematic forms.

The warning for admins should be along these lines:

> This form contains @count input elements, while PHP is configured to allow @limit. The max_input_vars limit should be increased to avoid this situation. See [link] for more information.

The warning for end users should be less technical:

> This form may be longer than this system can safely support. That results in it taking longer to be saved, and eventually possibly failing.

...or something like that.

**Comment by jenlampton** [comment]
> The warning should be shown after the form loads. When the error is shown (after the form has been submitted), the warning should be hidden. Otherwise it seems like unnecessary duplication.

I thought this too at first, but then after fighting with it for about 2 hours, I thought -- people both need to know that they just lost data, and that they are about to loose data again -- so, is it really unnecessary duplication?

Either way - I don't know if it's possible to prevent the warning after the error. I couldn't figure it out :)

> Showing both a success and an error message may be confusing.

Same here, this change might not be possible.

> I think that all users should see this same error after submitting the form

Which error? the admin one or the non-admin one?

> This form may be longer than this system can safely support. That results in it taking longer to be saved, and eventually possibly failing. Please notify the site administrator about this.

That is not what happens. (There's no change in time it takes to submit.)  The form submits as usual, but all the elements that are over the limit are not included when Backdrop gets the submitted values. Since the submit button is usually at the end of the form, it is often missing when the submit handler fires. If the submit handler depended on that button (many use the value to find out what to do: save? delete? cancel?) it is common that nothing happens at all -- but what will actually go wrong depends on the form, and how many values were missing by the time Backdrop gets the results. There's no way for us to know...  hence `may be too long to work properly`

**Comment by klonos** [comment]
> I thought this too at first, but then after fighting with it for about 2 hours....

Yeah, notice how I've edited my comment to remove that only seconds after submitting it? 😅

**Comment by klonos** [comment]
> Which error? the admin one or the non-admin one?

In case it wasn't clear from my comment, the ERROR should be a single thing for all users:

> Changes have been saved, but some data may have been lost. [ERROR]

**Comment by klonos** [comment]
...then we should show a different warning, to admins vs non-technical end users.

**Comment by klonos** [comment]
> That is not what happens. ...

I see, how about something like this then?:

> This form may be longer than this system can safely support, so some submitted data may not be sent through when saving.

**Comment by jenlampton** [comment]
It's awfully wordy. What's wrong with `This form may be too long to work properly`?

**Comment by jenlampton** [comment]
I came back to this issue after @docwilmot recommended a fix for the double message in the preview issue, but now, after looking at the screenshot below, I'm not sure any of the messages need to be removed. Thoughts?

<img width="1322" alt="Screen Shot 2020-12-11 at 11 11 57 AM" src="https://user-images.githubusercontent.com/397895/101944539-bb1dd500-3ba1-11eb-9ee9-e556e8d72b63.png">

**Comment by BWPanda** [comment]
@jenlampton Can you not combine the green and the red messages, since they both seem to be talking about the same thing (i.e. the form that was just submitted)? Something like:

> The changes have been submitted, but since the form was longer than the system can safely support, not all changes may have been successfully saved.

I'm not sure what colour you'd use for this message though...

**Comment by BWPanda** [comment]
So workflow would be something like:

- Open a form; yellow warning displayed
- Make changes and submit form; hybrid green/red 'submit' message displayed (see wording in comment above), yellow warning hidden
- Make more changes; same as last step - green/red 'submit' message displayed, no yellow warning
- Navigate away then come back, or just refresh form; only yellow warning displayed

**Comment by jenlampton** [comment]
Okay, let me give that a try.

**Comment by jenlampton** [comment]
Well, It looks like the recommendation of checking `$_SESSION['messages']` for a specific message may not work perfectly here, since the success message could be nearly any string, because this error could happen on nearly any form. 

We have no way of knowing any specific form's success message, but I can special-case some core admin forms that we know to be problematic. new PR up for review:

**Comment by klonos** [comment]
I haven't tested, but the last iteration of screenshots seems like a proper workflow to me 👍 

One thing is that, as @BWPanda suggested, the error message should say that the form was submitted. So perhaps change `Some data may have been lost.` to `The changes have been submitted, but some data may have been lost.` ...also, perhaps split the text into 2 bullet points within the error:

> - The changes have been submitted, but some data may have been lost.
> - The form contains ...

**Comment by jenlampton** [comment]
> perhaps split the text into 2 bullet points within the error:

So, set two separate messages then? (this is what determines the bullets) I'll give that a try...

**Comment by jenlampton** [comment]
Rebased and pushed a more recent PR. This could use another review with fresh eyes :)

**Comment by BWPanda** [comment]
Left a review on the PR requesting various changes.

**Comment by jenlampton** [comment]
I've merged a ton of recommended improvements (thanks @BWPanda  and @klonos!) and addressed a few comments. New PR up for review!

**Comment by BWPanda** [comment]
Looks good to me.

**Comment by klonos** [comment]
@jenlampton things look really good now, but tests are failing.

**Comment by jenlampton** [comment]
okay, I've rebased and pushed a new PR. If tests are still failing I'll dig back in deeper :)

**Comment by yorkshire-pudding** [comment]
I've just been hit by this again; a silent error where the node edit form just stops working while editing paragraphs I was experimenting with the PHP memory to see if that was the issue; spotted the `max_input_vars` and remembers this issue.

Still looks like tests are failing for some reason on https://github.com/backdrop/backdrop/pull/2955#issuecomment-1132395104 but it would be really good to get this in.

**Comment by argiepiano** [comment]
Probably a good idea to rebase and rerun tests?

**Comment by jenlampton** [comment]
Ive rebased and am rerunning tests now. Will check the results after the meeting :)

**Comment by jenlampton** [comment]
Today in the weekly meeting @quicksketch mentioned that maybe we should split the status report update out of this PR, and handle it on its own, so I have created https://github.com/backdrop/backdrop-issues/issues/5771. 

The status report message could get into 1.23.0, and would be a benefit on its own. Then we can keep working on the error messaging here, and revise as needed before merging.

**Comment by argiepiano** [comment]
Although the tests seem to have passed, there are some silent failures. Look at several of the Locale tests in 7.4 3/3 and 8.1 3/3 and click on the "Run tests" accordion. 

While the tests say that they've passed, there is extraneous HTML output to that shows errors and include stuff like ` <h1>Error</h1> syntax error, unexpected variable &quot;$this&quot;, expecting &quot;)&quot;    </main>
  </div> `

I don't know if these are produced by this PR or something else.

**Comment by indigoxela** [comment]
> syntax error, unexpected variable "$this", expecting ")"

@argiepiano  That's a [known regression](https://github.com/backdrop/backdrop-issues/issues/5764), unrelated to this PR.
